onyo/lib/onyo.py

```diff
@@ -906,6 +930,227 @@ def valid_name(self, asset: Path) -> bool:
+                if depth is not None and depth == 0:
+                    continue
+                # before max depth is reached, go down one level more
+                elif depth is not None and depth > 0:
+                    assets += self._select_assets_from_directory(path, depth - 1, recursive)
+                # flag is not set, so use infinite depth
+                elif not depth:
+                    assets += self._select_assets_from_directory(path, depth, recursive)
+
+        return assets
+
+    def _diff_changes(self) -> str:
+        """
+        Return a diff of all uncommitted changes. The format is a simplified
+        version of `git diff`.
+        """
+        diff = self._git(['--no-pager', 'diff', 'HEAD']).splitlines()
+        diff = [line.strip().replace("+++ b/", "\n").replace("+++ /dev/null", "\n")
+                for line in diff if len(line) > 0 and line[0] in ['+', '-'] and not
+                line[0:4] == '--- ' or "rename" in line]
+
+        return "\n".join(diff).strip()
+
+    def _read_asset(self, asset: Path) -> dict:
+        """
+        Read and return the contents of an asset as a dictionary.
+        """
+        yaml = YAML(typ='rt', pure=True)
+        contents = dict()
+        try:
+            contents = yaml.load(asset)
+        except scanner.ScannerError as e:
+            print(e)
+        if contents is None:
+            return dict()
+
+        return contents
+
+    def _set_sanitize(self, paths: Iterable[Union[Path, str]], recursive: bool,
+                      depth: Union[int, None]) -> list[Path]:
+        """
+        Check and normalize a list of paths. Select all assets from `paths` and
+        return them as a list of absolute Paths.
+        """
+        error_path_absent = []
+        error_path_protected = []
+        paths_to_set = []
+
+        if depth and depth < 0:
+            log.error(f"--depth must be positive, but is {depth}.")
+            raise ValueError(f"--depth must positive, but is {depth}.")
+
+        if depth and not recursive:
+            log.error("The flag --depth requires --recursive.")
+            raise ValueError("The flag --depth requires --recursive.")
+
+        for p in paths:
+
+            full_path = Path(self.opdir, p).resolve()
+
+            # paths must exist
+            if not full_path.exists():
+                error_path_absent.append(p)
+                continue
+
+            # protected paths
+            if self._is_protected_path(full_path):
+                error_path_protected.append(p)
+                continue
+
+            if full_path.is_dir():
+                paths_to_set += self._select_assets_from_directory(full_path, depth, recursive)
+            else:
+                paths_to_set.append(full_path)
+
+        if error_path_absent:
+            log.error('The following paths do not exist:\n' +
+                      '\n'.join(error_path_absent) + '\n' +
```

Comment:
This will fail on Paths. I recommend using `self._n_join`. It'll take care of casting and iterating, and allow you to avoid quite a few bugs.